src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -99,17 +99,49 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
 
             case 2:
                 genConsumeOperands(node);
+                op1Reg = op1->gtRegNum;
+                op2Reg = op2->gtRegNum;
                 if (category == HW_Category_MemoryStore)
                 {
-                    emit->emitIns_AR_R(ins, simdSize, op2->gtRegNum, op1->gtRegNum, 0);
+                    emit->emitIns_AR_R(ins, simdSize, op2Reg, op1Reg, 0);
                 }
                 else if ((ival != -1) && varTypeIsFloating(baseType))
                 {
                     genHWIntrinsic_R_R_RM_I(node, ins);
                 }
                 else if (category == HW_Category_MemoryLoad)
                 {
-                    emit->emitIns_SIMD_R_R_AR(ins, simdSize, targetReg, op1->gtRegNum, op2->gtRegNum);
+                    emit->emitIns_SIMD_R_R_AR(ins, simdSize, targetReg, op1Reg, op2Reg);
+                }
+                else if (Compiler::isImmHWIntrinsic(intrinsicID, op2))
+                {
+                    auto emitSwCase = [&](int i) {
+                        if ((ins == INS_psrldq || ins == INS_pslldq) && i > 15)
+                        {
+                            // TODO-XARCH-BUG: the emitter cannot work with psrldq/pslldq on i > 127.
+                            //                 Use a same-semantic alternative as the workaround
+                            emit->emitIns_SIMD_R_R_R(INS_xorps, simdSize, targetReg, targetReg, targetReg);
```

Comment:
@CarolEidt @tannergooding The emitter cannot work with `psrldq/pslldq` on `imm8` > 127. So I made this workaround and logged this bug at https://github.com/dotnet/coreclr/issues/16543. 
Please see https://github.com/dotnet/coreclr/pull/16183#discussion_r170104457 for more contexts.

Reply:
That is, I presume the problem is that we are coverting a `unsigned int8` to a `signed native int`, which will zero extend (255 -> 255), we want it to sign-extend instead so we would need to convert `unsigned int8` to `signed int8` (255 -> -1) and then to `signed native int` so it will sign extend (-1 -> -1)

Reply:
> Can we not convert to unsigned char and then to char in order to ensure that "must be signed" semantics work?

No, I tried. And other code relies on the current behavior. Additionally, this bug is not only related to “valInByte”, 32-bit build has more problems...

Reply:
Could you elaborate?

The `valInByte` check should succeed if you add a cast in the `emitIns_SIMD_*_I` methods, which should unblock the assert/etc.

What code is relying on the "current behavior" (I assume this means the zero-extension) if you do cast the value (immediate values for SIMD instructions should always be 8-bit)?

What other problems does the 32-bit build have?

Reply:
> Can we not convert to unsigned char and then to char in order to ensure that "must be signed" semantics work?

No amount of casting will make a value greater than 127 fit in a signed imm8...

> TODO-XARCH-BUG

I'm not sure why is this a bug. Emitting `xorps` is fine. Emitting `pxor` is also an option if you're concerned about data bypass delays (though I'm not sure if that applies to xor and anyway it doesn't really matter). Emitting `psrldq reg, 16` is yet another option, if you really, really want to emit a shift instructions (though I have no idea why would you want that) or if you want to keep the code shorter and simpler (now that would be a good reason).

Reply:
> if you really, really want to emit a shift instructions

Not really, actually I want to replace most of the shift instructions by XOR if possible. Please see  https://github.com/dotnet/coreclr/pull/16183/files#diff-46e84d907ca2183faf22c1345e4e04aeR588

Reply:
Not sure if it's worth doing that all other the place, it seems unlikely that people will write `shift(x, large constant shift count)` on purpose. This is more likely to be useful in C++, where extensive use of template metaprogramming can lead to cases where a large shift count magically appears and people even expect it to work well.

It's IMO worth doing if it cuts down complexity in common JIT code paths. As in this case, why bother making `emitIns_R_I` more complex when you can just emit `xor`?

Last time when I tried to add a bunch of new special cases to the emitter I ended up with a 0.2% regression in instruction count and rolled back most of the changes...

Reply:
>  This is more likely to be useful in C++, where extensive use of template metaprogramming can lead to cases where a large shift count magically appears and people even expect it to work well.

> Last time when I tried to add a bunch of new special cases to the emitter I ended up with a 0.2% regression in instruction count and rolled back most of the changes...

Thank you so much for teaching! Emitting `xor` for `psrldq/pslldq` looks like a good solution now.

Reply:
> No amount of casting will make a value greater than 127 fit in a signed imm8...

Actually it should probably work. It's implementation behavior in C++ but then the emitter does it anyway so it the C++ compiler doesn't like it then we'd have bigger problems anyway.

Doesn't really matter as one way or another you need special casing to handle this, either here or in the emitter. I'd move this special casing to the newly added `emitIns_SIMD_R_R_I` that too has special casing for shifts.